internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -0,0 +1,755 @@
+	databaseSecretName                  = databaseName
+	databasePort                  int32 = 5432
+	servicePort                   int32 = 8090
+	assistedServiceDeploymentName       = "assisted-service"
+	postgresDeploymentName              = databaseName
+
+	// assistedServiceContainerName is the Name property of the assisted-service container
+	assistedServiceContainerName string = "assisted-service"
+	// databaseContainerName is the Name property of the postgres container
+	databaseContainerName string = databaseName
+)
+
+// AgentServiceConfigReconciler reconciles a AgentServiceConfig object
+type AgentServiceConfigReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	// the Namespace the operator is running in
+	// TODO(djzager): This must be configured on the operator deployment (ie. in the CSV
+	// as from the metadata.Namespace
+	Namespace string
+}
+
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs/finalizers,verbs=update
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=route.openshift.io,resources=routes,verbs=get;list;watch;create;update;patch;delete
+
+func (r *AgentServiceConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	_ = r.Log.WithValues("agentserviceconfig", req.NamespacedName)
+
+	instance := &adiiov1alpha1.AgentServiceConfig{}
+
+	// We only support one AgentServiceConfig per cluster, and it must be called "hive". This prevents installing
+	// AgentService more than once in the cluster.
+	if req.NamespacedName.Name != agentServiceConfigName {
+		err := fmt.Errorf("Invalid name (%s) expected %s", req.NamespacedName.Name, agentServiceConfigName)
+		r.Log.Error(err, fmt.Sprintf("Only one AgentServiceConfig supported per cluster and must be named '%s'", agentServiceConfigName))
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.Get(ctx, req.NamespacedName, instance); err != nil {
+		r.Log.Error(err, "Failed to get resource", req.NamespacedName)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	for _, f := range []func(context.Context, *adiiov1alpha1.AgentServiceConfig) error{
+		r.ensureFilesystemStorage,
+		r.ensureDatabaseStorage,
+		r.ensureAgentService,
+		r.ensurePostgresService,
+		r.ensureAgentRoute,
+		r.ensurePostgresSecret,
+		r.ensurePostgresDeployment,
```

Comment:
Maybe now isn't the best time to ask the question, but should we just run postgres as another container in the same pod as the assisted service? They'll both always be singletons.

Reply:
Thinking more about it, now's definitely the time! ;) If they're in different pods, we need to secure the network traffic between them. So we'd need to set up the DB with a TLS cert pair and the client side with a matching CA. It would be easier to just run them in the same pod and talk on localhost, unless this has already been accounted for somehow.

Reply:
Leaving this as unresolved but I think this is in the same class of issues as validating the storageClass. What do you think?

Reply:
Not sure if I forgot to refresh the page or totally missed your second comment before writing :point_up: .

Fair points. I'll give it a shot.